lib/utils/platform/iot_clock_afr.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2018 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file iot_clock_afr.c
+ * @brief Implementation of the functions in iot_clock.h for Amazon FreeRTOS systems.
+ */
+
+/* Build using a config header, if provided. */
```

Comment:
It's possible to define everything with a compiler command line option like `-D` (though not recommended).  And I think some people may want to build with the defaults (libraries should provide default values for undefined configuration constants).  Though it might not make sense for the platform abstractions.

Reply:
If we were careful and ONLY had defines in the config file this would be true, but we have let some type definitions sneak into the config files. It is not possible to define types using -D, so that is not quite true any longer... 

Platform types are now included from the header and I spotted some other types in there as well. We should fix this either way.

Reply:
Here is the exemplification of the two cases:
files i.h is the platform independent file (e.g. the one that define IotMutex_t as ```typedef _IotSystemMutex_t       IotMutex_t;```)
file d.h is the one that defines the implementation, e.g. the one that typedefs 
```typedef pthread_mutex_t   _IotSystemMutex_t;```)
then there is a config.h.

with the current implementation, i.h needs to include config.h, which includes d.h.
pros: i.h and d.h an be documented independently. 
cons: one ore intermediate type.

with the alternate proposal, config.h would need to typedef those types directly
pro: one less type
cons: cannot do (or more difficult) doxygen docs directly in code